applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+
+#include <device.h>
+#include <gpio.h>
+
+#include "selector_hw_def.h"
+
+#include "event_manager.h"
+#include "selector_event.h"
+#include "power_event.h"
+
+#define MODULE selector_hw
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_SELECTOR_HW_LOG_LEVEL);
+
+#define DEFAULT_POSITION UCHAR_MAX
+#define DEBOUNCE_DELAY K_MSEC(5)
+
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_OFF,
+	STATE_ERROR,
+};
+
+struct selector {
+	const struct selector_config *config;
+	struct gpio_callback gpio_cb[ARRAY_SIZE(port_map)];
+	struct k_delayed_work work;
+	u8_t position;
+};
+
+static struct device *gpio_dev[ARRAY_SIZE(port_map)];
+static struct selector selectors[ARRAY_SIZE(selector_config)];
+static enum state state;
+
+
+static void selector_event_send(const struct selector *selector)
+{
+	__ASSERT_NO_MSG((selector->position < selector->config->pins_size) ||
+			(selector->position == DEFAULT_POSITION));
+
+	struct selector_event *event = new_selector_event();
+
+	event->selector_id = selector->config->id;
+	event->position = selector->position;
+
+	EVENT_SUBMIT(event);
+}
+
+static int read_state(struct selector *selector)
+{
+	u32_t value = 0;
+	const struct gpio_pin *sel_pins = selector->config->pins;
+	int err = 0;
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		err = gpio_pin_read(gpio_dev[sel_pins[i].port],
+				    sel_pins[i].pin,
+				    &value);
+
+		if (err) {
+			LOG_ERR("Cannot read value from port:%" PRIu8
+				" pin: %" PRIu8, sel_pins[i].port,
+				sel_pins[i].pin);
+			break;
+		}
+
+		if (value) {
+			selector->position = i;
+			break;
+		}
+	}
+
+	if (err) {
+		return err;
+	}
+
+	if (!value) {
+		selector->position = DEFAULT_POSITION;
+		LOG_WRN("Selector %" PRIu8 " state is unknown",
+			selector->config->id);
+	}
+
+	selector_event_send(selector);
+
+	return 0;
+}
+
+static int enable_interrupts_lock(struct selector *selector)
+{
+	const struct gpio_pin *sel_pins = selector->config->pins;
+	int key = irq_lock();
+	int err = 0;
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		if (i == selector->position) {
+			continue;
+		}
+
+		err = gpio_pin_enable_callback(gpio_dev[sel_pins[i].port],
```

Comment:
since we want to see if switched moved from current position, wouldn't it be easier to hook the interrupt to the currently selected pin? You would have to reconfigure isr to wait for level change but then we would enable only one callback

Reply:
this would require additional code for checks when switch is in undefined state - discussed offline